File: src/platform_context_graph/data_intelligence/dbt_sql_expressions.py
"""Expression-shape helpers for compiled dbt SQL lineage."""

from __future__ import annotations

import re

from .dbt_sql_identifiers import unqualified_identifiers

_BARE_IDENTIFIER_RE = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")
_QUALIFIED_REFERENCE_RE = re.compile(
    r"^[A-Za-z_][A-Za-z0-9_]*\.(?:\*|[A-Za-z_][A-Za-z0-9_]*)$"
)
_QUALIFIED_REFERENCE_SCAN_RE = re.compile(
    r"\b(?P<alias>[A-Za-z_][A-Za-z0-9_]*)\."
    r"(?P<column>[A-Za-z_][A-Za-z0-9_]*)(?=[^A-Za-z0-9_]|$)"
)
_FUNCTION_CALL_RE = re.compile(
    r"^(?P<name>[A-Za-z_][A-Za-z0-9_]*)\((?P<arguments>.*)\)$",
    re.DOTALL,
)
_FUNCTION_CALL_SCAN_RE = re.compile(
    r"\b(?P<name>[A-Za-z_][A-Za-z0-9_]*)\s*\(",
)
_SINGLE_QUOTED_LITERAL_RE = re.compile(r"^'(?:[^'\\\\]|\\\\.)*'$", re.DOTALL)
_SINGLE_QUOTED_LITERAL_SCAN_RE = re.compile(r"'(?:[^'\\\\]|\\\\.)*'", re.DOTALL)
_NUMERIC_LITERAL_RE = re.compile(r"^[+-]?(?:\d+(?:\.\d+)?|\.\d+)$")
_NUMERIC_LITERAL_SCAN_RE = re.compile(
    r"(?<![A-Za-z0-9_])(?:[+-]?(?:\d+(?:\.\d+)?|\.\d+))(?![A-Za-z0-9_])"
)
_TYPE_IDENTIFIER_RE = re.compile(r"\b[A-Za-z_][A-Za-z0-9_]*\b")
_CASE_EXPRESSION_RE = re.compile(r"^case\b.*\bend$", re.IGNORECASE | re.DOTALL)
_CASE_KEYWORD_RE = re.compile(
    r"\b(?:case|when|then|else|end|is|null|and|or|not|in|like|between|true|false)\b",
    re.IGNORECASE,
)
_AGGREGATE_EXPRESSION_REASON = "aggregate_expression_semantics_not_captured"
_DERIVED_EXPRESSION_REASON = "derived_expression_semantics_not_captured"
_MULTI_INPUT_EXPRESSION_REASON = "multi_input_expression_semantics_not_captured"
_AGGREGATE_FUNCTIONS = {
    "avg",
    "count",
    "max",
    "min",
    "sum",
}
_SIMPLE_SCALAR_FUNCTIONS = {
    "upper",
    "lower",
    "trim",
    "ltrim",
    "rtrim",
}
_LITERAL_PARAMETER_SCALAR_FUNCTIONS = {
    "date_trunc",
}


def expression_requires_partial_reporting(expression: str) -> bool:
    """Return whether a projection expression should report partial semantics."""

    return expression_partial_reason(expression) is not None


def expression_transform_metadata(expression: str) -> dict[str, str] | None:
    """Return transform metadata for supported non-identity expressions."""

    normalized = _strip_wrapping_parentheses(expression.strip())
    if not normalized:
        return None
    if _BARE_IDENTIFIER_RE.fullmatch(normalized):
        return None
    if _QUALIFIED_REFERENCE_RE.fullmatch(normalized):
        return None

    cast_expression = _supported_cast_expression(normalized)
    if cast_expression is not None:
        return {
            "transform_kind": "cast",
            "transform_expression": normalized,
        }
    if _is_supported_case_expression(normalized):
        return {
            "transform_kind": "case",
            "transform_expression": normalized,
        }
    if _is_supported_arithmetic_expression(normalized):
        return {
            "transform_kind": "arithmetic",
            "transform_expression": normalized,
        }

    match = _FUNCTION_CALL_RE.fullmatch(normalized)
    if match is None:
        return None

    function_name = match.group("name").strip().lower()
    arguments = _split_top_level_arguments(match.group("arguments"))
    if function_name in _SIMPLE_SCALAR_FUNCTIONS and len(arguments) == 1:
        if _is_simple_reference_expression(arguments[0]):
            return {
                "transform_kind": function_name,
                "transform_expression": normalized,
            }
        return None
    if function_name == "coalesce" and len(arguments) >= 2:
        if _is_simple_reference_expression(arguments[0]) and all(
            _is_literal_expression(argument) for argument in arguments[1:]
        ):
            return {
                "transform_kind": "coalesce",
                "transform_expression": normalized,
            }
        return None
    if function_name in _LITERAL_PARAMETER_SCALAR_FUNCTIONS and len(arguments) >= 2:
        reference_arguments = [
            argument
            for argument in arguments
            if _is_simple_reference_expression(argument)
        ]
        if len(reference_arguments) != 1:
            return None
        if all(
            _is_simple_reference_expression(argument)
            or _is_literal_expression(argument)
            for argument in arguments
        ):
            return {
                "transform_kind": function_name,
                "transform_expression": normalized,
            }
    return None


def expression_partial_reason(expression: str) -> str | None:
    """Return a specific partial-lineage reason when the expression is unsupported."""

    normalized = _strip_wrapping_parentheses(expression.strip())
    if not normalized:
        return None
    if _BARE_IDENTIFIER_RE.fullmatch(normalized):
        return None
    if _QUALIFIED_REFERENCE_RE.fullmatch(normalized):
        return None
    if _is_supported_case_expression(normalized):
        return None
    if _is_supported_arithmetic_expression(normalized):
        return None
    if _is_supported_scalar_wrapper(normalized):
        return None
    function_reason = _unsupported_function_reason(normalized)
    if function_reason is not None:
        return function_reason
    return _DERIVED_EXPRESSION_REASON


def expression_ignored_identifiers(expression: str) -> set[str]:
    """Return extra bare identifiers that should be ignored for this expression."""

    cast_expression = _supported_cast_expression(expression)
    if cast_expression is None:
        return set()
    return {
        match.group(0)
        for match in _TYPE_IDENTIFIER_RE.finditer(cast_expression[1])
    }


def derived_expression_gap(
    *,
    expression: str,
    model_name: str,
    reason: str,
) -> dict[str, str]:
    """Return the standardized unresolved-gap record for one derived expression."""

    return {
        "expression": expression.strip(),
        "model_name": model_name,
        "reason": reason,
    }


def _strip_wrapping_parentheses(expression: str) -> str:
    """Strip balanced outer parentheses from a projection expression."""

    normalized = expression.strip()
    while normalized.startswith("(") and normalized.endswith(")"):
        depth = 0
        balanced = True
        for index, character in enumerate(normalized):
            if character == "(":
                depth += 1
            elif character == ")":
                depth -= 1
                if depth == 0 and index != len(normalized) - 1:
                    balanced = False
                    break
        if not balanced or depth != 0:
            return normalized
        normalized = normalized[1:-1].strip()
    return normalized


def _is_supported_scalar_wrapper(expression: str) -> bool:
    """Return whether the expression is a supported one-column scalar wrapper."""

    if _supported_cast_expression(expression) is not None:
        return True

    match = _FUNCTION_CALL_RE.fullmatch(expression)
    if match is None:
        return False

    function_name = match.group("name").strip().lower()
    arguments = _split_top_level_arguments(match.group("arguments"))
    if function_name in _SIMPLE_SCALAR_FUNCTIONS and len(arguments) == 1:
        return _is_simple_reference_expression(arguments[0])
    if function_name == "coalesce" and len(arguments) >= 2:
        return _is_simple_reference_expression(arguments[0]) and all(
            _is_literal_expression(argument) for argument in arguments[1:]
        )
    if function_name in _LITERAL_PARAMETER_SCALAR_FUNCTIONS and len(arguments) >= 2:
        reference_arguments = [
            argument for argument in arguments if _is_simple_reference_expression(argument)
        ]
        if len(reference_arguments) != 1:
            return False
        return all(
            _is_simple_reference_expression(argument)
            or _is_literal_expression(argument)
            for argument in arguments
        )
    return False


def _is_supported_case_expression(expression: str) -> bool:
    """Return whether the expression is a supported one-column CASE transform."""

    if _CASE_EXPRESSION_RE.fullmatch(expression) is None:
        return False
    if _has_unsupported_function_call(expression):
        return False
    references = _simple_reference_tokens(expression)
    if len(references) != 1:
        return False

    collapsed = _collapsed_shape(
        expression,
        references=references,
        strip_case_keywords=True,
    )
    return re.fullmatch(r"[\s()=<>!,+\-*/%]*", collapsed) is not None


def _is_supported_arithmetic_expression(expression: str) -> bool:
    """Return whether the expression is a supported one-column arithmetic transform."""

    if not any(operator in expression for operator in ("+", "-", "*", "/", "%")):
        return False
    if _has_unsupported_function_call(expression):
        return False
    references = _simple_reference_tokens(expression)
    if len(references) != 1:
        return False

    collapsed = _collapsed_shape(expression, references=references)
    return re.fullmatch(r"[\s()+\-*/%]*", collapsed) is not None


def _supported_cast_expression(expression: str) -> tuple[str, str] | None:
    """Return the cast value/type when the CAST expression is supported."""

    match = _FUNCTION_CALL_RE.fullmatch(expression)
    if match is None or match.group("name").strip().lower() != "cast":
        return None

    value_expression, type_expression = _split_cast_arguments(match.group("arguments"))
    if value_expression is None or type_expression is None:
        return None
    if not _is_simple_reference_expression(value_expression):
        return None
    if not type_expression.strip():
        return None
    return value_expression, type_expression


def _unsupported_function_reason(expression: str) -> str | None:
    """Return a more specific unsupported-function reason when possible."""

    match = _FUNCTION_CALL_RE.fullmatch(expression)
    if match is None:
        return None

    function_name = match.group("name").strip().lower()
    if function_name in _AGGREGATE_FUNCTIONS:
        return _AGGREGATE_EXPRESSION_REASON

    arguments = _split_top_level_arguments(match.group("arguments"))
    reference_count = sum(
        1 for argument in arguments if _is_simple_reference_expression(argument)
    )
    if reference_count > 1:
        return _MULTI_INPUT_EXPRESSION_REASON
    return None


def _simple_reference_tokens(expression: str) -> tuple[str, ...]:
    """Return unique simple reference tokens used in one expression."""

    matched_identifiers: set[str] = set()
    tokens: list[str] = []
    seen_tokens: set[str] = set()
    for match in _QUALIFIED_REFERENCE_SCAN_RE.finditer(expression):
        token = match.group(0)
        matched_identifiers.add(match.group("alias"))
        matched_identifiers.add(match.group("column"))
        if token in seen_tokens:
            continue
        seen_tokens.add(token)
        tokens.append(token)
    for identifier in unqualified_identifiers(
        expression,
        matched_identifiers=matched_identifiers,
    ):
        if identifier in seen_tokens:
            continue
        seen_tokens.add(identifier)
        tokens.append(identifier)
    return tuple(tokens)


def _has_unsupported_function_call(expression: str) -> bool:
    """Return whether the expression contains a nested function call."""

    for match in _FUNCTION_CALL_SCAN_RE.finditer(expression):
        if match.group("name").strip().lower() in {
            "and",
            "case",
            "else",
            "end",
            "in",
            "is",
            "not",
            "or",
            "then",
            "when",
        }:
            continue
        return True
    return False


def _collapsed_shape(
    expression: str,
    *,
    references: tuple[str, ...],
    strip_case_keywords: bool = False,
) -> str:
    """Return one expression with references and literals reduced to shape tokens."""

    sanitized = _SINGLE_QUOTED_LITERAL_SCAN_RE.sub("0", expression)
    sanitized = _NUMERIC_LITERAL_SCAN_RE.sub("0", sanitized)
    sanitized = _replace_reference_tokens(sanitized, references)
    if strip_case_keywords:
        sanitized = _CASE_KEYWORD_RE.sub(" ", sanitized)
    return sanitized.replace("REF", "").replace("0", "")


def _replace_reference_tokens(expression: str, references: tuple[str, ...]) -> str:
    """Replace simple reference tokens with a uniform placeholder."""

    sanitized = expression
    for token in references:
        sanitized = re.sub(rf"\b{re.escape(token)}\b", "REF", sanitized)
    return sanitized


def _is_simple_reference_expression(expression: str) -> bool:
    """Return whether the expression is a direct supported column reference."""

    normalized = _strip_wrapping_parentheses(expression.strip())
    if not normalized:
        return False
    return bool(
        _BARE_IDENTIFIER_RE.fullmatch(normalized)
        or _QUALIFIED_REFERENCE_RE.fullmatch(normalized)
    )


def _is_literal_expression(expression: str) -> bool:
    """Return whether the expression is a supported literal fallback."""

    normalized = _strip_wrapping_parentheses(expression.strip())
    if not normalized:
        return False
    lowered = normalized.lower()
    if lowered in {"null", "true", "false"}:
        return True
    if _SINGLE_QUOTED_LITERAL_RE.fullmatch(normalized):
        return True
    return bool(_NUMERIC_LITERAL_RE.fullmatch(normalized))


def _split_top_level_arguments(arguments: str) -> list[str]:
    """Split one function-argument list on top-level commas."""

    items: list[str] = []
    current: list[str] = []
    depth = 0
    in_single_quote = False

    for character in arguments:
        if character == "'" and (not current or current[-1] != "\\"):
            in_single_quote = not in_single_quote
        elif not in_single_quote:
            if character == "(":
                depth += 1
            elif character == ")" and depth > 0:
                depth -= 1
            elif character == "," and depth == 0:
                item = "".join(current).strip()
                if item:
                    items.append(item)
                current = []
                continue
        current.append(character)

    tail = "".join(current).strip()
    if tail:
        items.append(tail)
    return items


def _split_cast_arguments(arguments: str) -> tuple[str | None, str | None]:
    """Split one CAST body into value and type expressions."""

    depth = 0
    in_single_quote = False
    lower_arguments = arguments.lower()

    for index, character in enumerate(arguments):
        if character == "'" and (index == 0 or arguments[index - 1] != "\\"):
            in_single_quote = not in_single_quote
            continue
        if in_single_quote:
            continue
        if character == "(":
            depth += 1
            continue
        if character == ")" and depth > 0:
            depth -= 1
            continue
        if depth != 0:
            continue
        if lower_arguments[index : index + 4] != " as ":
            continue
        value_expression = arguments[:index].strip()
        type_expression = arguments[index + 4 :].strip()
        return value_expression, type_expression
    return None, None


__all__ = [
    "derived_expression_gap",
    "expression_ignored_identifiers",
    "expression_partial_reason",
    "expression_transform_metadata",
    "expression_requires_partial_reporting",
]

File: go/internal/parser/dbt_sql_expressions.go
package parser

import (
	"regexp"
	"strings"
)

var (
	dbtBareIdentifierRe         = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)
	dbtQualifiedReferenceRe     = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*\.(?:\*|[A-Za-z_][A-Za-z0-9_]*)$`)
	dbtQualifiedReferenceScanRe = regexp.MustCompile(`\b(?P<alias>[A-Za-z_][A-Za-z0-9_]*)\.(?P<column>\*|[A-Za-z_][A-Za-z0-9_]*)`)
	dbtFunctionCallRe           = regexp.MustCompile(`^(?P<name>[A-Za-z_][A-Za-z0-9_]*)\((?P<arguments>.*)\)$`)
	dbtFunctionCallScanRe       = regexp.MustCompile(`\b(?P<name>[A-Za-z_][A-Za-z0-9_]*)\s*\(`)
	dbtWindowFunctionRe         = regexp.MustCompile(`(?is)^(?P<name>[A-Za-z_][A-Za-z0-9_]*)\((?P<arguments>.*)\)\s+over\s*\((?P<window>.*)\)$`)
	dbtSingleQuotedLiteralRe    = regexp.MustCompile(`(?s)^'(?:[^'\\]|\\.)*'$`)
	dbtSingleQuotedLiteralScan  = regexp.MustCompile(`(?s)'(?:[^'\\]|\\.)*'`)
	dbtNumericLiteralRe         = regexp.MustCompile(`^[+-]?(?:\d+(?:\.\d+)?|\.\d+)$`)
	dbtNumericLiteralScan       = regexp.MustCompile(`(^|[^A-Za-z0-9_])([+-]?(?:\d+(?:\.\d+)?|\.\d+))($|[^A-Za-z0-9_])`)
	dbtTypeIdentifierRe         = regexp.MustCompile(`\b[A-Za-z_][A-Za-z0-9_]*\b`)
	dbtCaseExpressionRe         = regexp.MustCompile(`(?is)^case\b.*\bend$`)
	dbtCaseKeywordRe            = regexp.MustCompile(`(?i)\b(?:case|when|then|else|end|is|null|and|or|not|in|like|between|true|false)\b`)
	dbtQualifiedMacroCallRe     = regexp.MustCompile(`^(?P<name>[A-Za-z_][A-Za-z0-9_]*(?:\.[A-Za-z_][A-Za-z0-9_]*)+)\((?P<arguments>.*)\)$`)
)

var (
	dbtAggregateFunctions          = map[string]struct{}{"avg": {}, "count": {}, "max": {}, "min": {}, "sum": {}}
	dbtSimpleScalarFunctions       = map[string]struct{}{"md5": {}, "upper": {}, "lower": {}, "trim": {}, "ltrim": {}, "rtrim": {}}
	dbtLiteralParameterScalarFuncs = map[string]struct{}{"date_trunc": {}}
	dbtMultiInputRowLevelFunctions = map[string]struct{}{"concat": {}}
)

const (
	dbtAggregateExpressionReason = "aggregate_expression_semantics_not_captured"
	dbtDerivedExpressionReason   = "derived_expression_semantics_not_captured"
	dbtMultiInputExpression      = "multi_input_expression_semantics_not_captured"
	dbtMacroExpressionReason     = "macro_expression_not_resolved"
	dbtTemplatedExpressionReason = "templated_expression_not_resolved"
	dbtWindowExpressionReason    = "window_expression_semantics_not_captured"
)

func expressionHonestyGapReason(expression string) string {
	normalized := stripWrappingParentheses(strings.TrimSpace(expression))
	if normalized == "" {
		return ""
	}
	if inner, ok := unwrapTemplatedExpression(normalized); ok && !dbtBareIdentifierRe.MatchString(inner) {
		if expressionPartialReason(inner) != "" {
			return dbtTemplatedExpressionReason
		}
		normalized = inner
	}
	if strings.Contains(normalized, "{{") || strings.Contains(normalized, "}}") || strings.Contains(normalized, "{%") || strings.Contains(normalized, "%}") {
		return dbtTemplatedExpressionReason
	}
	if dbtQualifiedMacroCallRe.MatchString(normalized) && !macroExpressionHasLineage(normalized) && !isSupportedQualifiedMacroExpression(normalized) {
		return dbtMacroExpressionReason
	}
	return ""
}

func expressionIgnoredIdentifiers(expression string) map[string]struct{} {
	result := make(map[string]struct{})
	valueExpression, typeExpression, ok := supportedCastExpression(expression)
	if !ok {
		return result
	}
	_ = valueExpression
	for _, match := range dbtTypeIdentifierRe.FindAllString(typeExpression, -1) {
		result[match] = struct{}{}
	}
	return result
}

func derivedExpressionGap(expression string, modelName string, reason string) map[string]string {
	return map[string]string{
		"expression": strings.TrimSpace(expression),
		"model_name": modelName,
		"reason":     reason,
	}
}

func expressionPartialReason(expression string) string {
	normalized := stripWrappingParentheses(strings.TrimSpace(expression))
	if normalized == "" {
		return ""
	}
	if inner, ok := unwrapTemplatedExpression(normalized); ok && !dbtBareIdentifierRe.MatchString(inner) {
		if expressionPartialReason(inner) != "" {
			return dbtTemplatedExpressionReason
		}
		normalized = inner
	}
	if reason := expressionHonestyGapReason(normalized); reason != "" {
		return reason
	}
	if dbtQualifiedMacroCallRe.MatchString(normalized) && macroExpressionHasLineage(normalized) {
		return ""
	}
	if dbtBareIdentifierRe.MatchString(normalized) || dbtQualifiedReferenceRe.MatchString(normalized) {
		return ""
	}
	if isSupportedCaseExpression(normalized) || isSupportedArithmeticExpression(normalized) || isSupportedScalarWrapper(normalized) {
		return ""
	}
	if isSupportedQualifiedMacroExpression(normalized) {
		return ""
	}
	if isSupportedAggregateExpression(normalized) {
		return ""
	}
	if isSupportedWindowExpression(normalized) {
		return ""
	}
	if reason := unsupportedFunctionReason(normalized); reason != "" {
		return reason
	}
	return dbtDerivedExpressionReason
}

func macroExpressionHasLineage(expression string) bool {
	matches := dbtQualifiedMacroCallRe.FindStringSubmatch(strings.TrimSpace(expression))
	if matches == nil {
		return false
	}
	return len(simpleReferenceTokens(matches[2])) > 0
}

func isSupportedAggregateExpression(expression string) bool {
	matches := dbtFunctionCallRe.FindStringSubmatch(expression)
	if matches == nil {
		return false
	}
	functionName := strings.ToLower(strings.TrimSpace(matches[1]))
	if _, ok := dbtAggregateFunctions[functionName]; !ok {
		return false
	}
	return supportsRowLevelArguments(splitTopLevelArguments(matches[2]))
}

func isSupportedWindowExpression(expression string) bool {
	matches := dbtWindowFunctionRe.FindStringSubmatch(expression)
	if matches == nil {
		return false
	}
	functionName := strings.ToLower(strings.TrimSpace(matches[1]))
	if _, ok := dbtAggregateFunctions[functionName]; !ok {
		return false
	}
	if !supportsRowLevelArguments(splitTopLevelArguments(matches[2])) {
		return false
	}
	return len(simpleReferenceTokens(expression)) > 0
}

func isSupportedQualifiedMacroExpression(expression string) bool {
	matches := dbtQualifiedMacroCallRe.FindStringSubmatch(expression)
	if matches == nil {
		return false
	}
	return supportsRowLevelArguments(splitTopLevelArguments(matches[2])) && len(simpleReferenceTokens(matches[2])) > 0
}

func expressionTransformMetadata(expression string) map[string]string {
	normalized := stripWrappingParentheses(strings.TrimSpace(expression))
	if normalized == "" || dbtBareIdentifierRe.MatchString(normalized) || dbtQualifiedReferenceRe.MatchString(normalized) {
		return nil
	}
	if expressionHonestyGapReason(normalized) != "" {
		return nil
	}
	if _, _, ok := supportedCastExpression(normalized); ok {
		return map[string]string{"transform_kind": "cast", "transform_expression": normalized}
	}
	if isSupportedCaseExpression(normalized) {
		return map[string]string{"transform_kind": "case", "transform_expression": normalized}
	}
	if isSupportedArithmeticExpression(normalized) {
		return map[string]string{"transform_kind": "arithmetic", "transform_expression": normalized}
	}
	if metadata := partialTransformMetadata(normalized); metadata != nil {
		return metadata
	}
	return supportedFunctionMetadata(normalized)
}

func stripWrappingParentheses(expression string) string {
	normalized := strings.TrimSpace(expression)
	for strings.HasPrefix(normalized, "(") && strings.HasSuffix(normalized, ")") {
		depth := 0
		balanced := true
		for index, character := range normalized {
			switch character {
			case '(':
				depth++
			case ')':
				depth--
				if depth == 0 && index != len(normalized)-1 {
					balanced = false
				}
			}
			if !balanced {
				break
			}
		}
		if !balanced || depth != 0 {
			return normalized
		}
		normalized = strings.TrimSpace(normalized[1 : len(normalized)-1])
	}
	return normalized
}

func isSupportedScalarWrapper(expression string) bool {
	_, _, ok := supportedCastExpression(expression)
	return ok || supportedFunctionMetadata(expression) != nil
}

func isSupportedRowLevelExpression(expression string) bool {
	normalized := stripWrappingParentheses(strings.TrimSpace(expression))
	if normalized == "" {
		return false
	}
	return isSimpleReferenceExpression(normalized) ||
		isLiteralExpression(normalized) ||
		isSupportedScalarWrapper(normalized) ||
		isSupportedQualifiedMacroExpression(normalized)
}

func isSupportedCaseExpression(expression string) bool {
	if !dbtCaseExpressionRe.MatchString(expression) || hasUnsupportedFunctionCall(expression) {
		return false
	}
	references := simpleReferenceTokens(expression)
	if len(references) == 0 {
		return false
	}
	collapsed := collapsedShape(expression, references, true)
	return regexp.MustCompile(`^[\s()=<>!,+\-*/%]*$`).MatchString(collapsed)
}

func isSupportedArithmeticExpression(expression string) bool {
	if !strings.ContainsAny(expression, "+-*/%") || hasUnsupportedFunctionCall(expression) {
		return false
	}
	references := simpleReferenceTokens(expression)
	if len(references) == 0 {
		return false
	}
	collapsed := collapsedShape(expression, references, false)
	return regexp.MustCompile(`^[\s()+\-*/%]*$`).MatchString(collapsed)
}

func supportedCastExpression(expression string) (string, string, bool) {
	matches := dbtFunctionCallRe.FindStringSubmatch(expression)
	if matches == nil || strings.ToLower(strings.TrimSpace(matches[1])) != "cast" {
		return "", "", false
	}
	valueExpression, typeExpression, ok := splitCastArguments(matches[2])
	if !ok || !isSupportedRowLevelExpression(valueExpression) || strings.TrimSpace(typeExpression) == "" {
		return "", "", false
	}
	return valueExpression, typeExpression, true
}

func unsupportedFunctionReason(expression string) string {
	if matches := dbtWindowFunctionRe.FindStringSubmatch(expression); matches != nil {
		if isSupportedWindowExpression(expression) {
			return ""
		}
		return dbtWindowExpressionReason
	}
	matches := dbtFunctionCallRe.FindStringSubmatch(expression)
	if matches == nil || supportedFunctionMetadata(expression) != nil {
		return ""
	}
	functionName := strings.ToLower(strings.TrimSpace(matches[1]))
	if _, ok := dbtAggregateFunctions[functionName]; ok {
		return dbtAggregateExpressionReason
	}
	arguments := splitTopLevelArguments(matches[2])
	referenceCount := 0
	for _, argument := range arguments {
		if isSimpleReferenceExpression(argument) {
			referenceCount++
		}
	}
	if referenceCount > 1 {
		return dbtMultiInputExpression
	}
	return ""
}

func partialTransformMetadata(expression string) map[string]string {
	if matches := dbtWindowFunctionRe.FindStringSubmatch(expression); matches != nil && len(simpleReferenceTokens(expression)) > 0 {
		return map[string]string{
			"transform_kind":       "window_" + strings.ToLower(strings.TrimSpace(matches[1])),
			"transform_expression": expression,
		}
	}
	matches := dbtFunctionCallRe.FindStringSubmatch(expression)
	if matches == nil {
		return nil
	}
	functionName := strings.ToLower(strings.TrimSpace(matches[1]))
	if _, ok := dbtAggregateFunctions[functionName]; ok && supportsRowLevelArguments(splitTopLevelArguments(matches[2])) {
		return map[string]string{"transform_kind": functionName, "transform_expression": expression}
	}
	return nil
}

func supportedFunctionMetadata(expression string) map[string]string {
	matches := dbtFunctionCallRe.FindStringSubmatch(expression)
	if matches == nil {
		return nil
	}
	functionName := strings.ToLower(strings.TrimSpace(matches[1]))
	arguments := splitTopLevelArguments(matches[2])
	if _, ok := dbtSimpleScalarFunctions[functionName]; ok && len(arguments) == 1 && isSupportedRowLevelExpression(arguments[0]) {
		return map[string]string{"transform_kind": functionName, "transform_expression": expression}
	}
	if functionName == "coalesce" && len(arguments) >= 2 && supportsRowLevelArguments(arguments) {
		return map[string]string{"transform_kind": "coalesce", "transform_expression": expression}
	}
	if _, ok := dbtLiteralParameterScalarFuncs[functionName]; ok && len(arguments) >= 2 {
		referenceArguments := 0
		valid := true
		for _, argument := range arguments {
			if isSimpleReferenceExpression(argument) {
				referenceArguments++
				continue
			}
			if !isLiteralExpression(argument) {
				valid = false
			}
		}
		if valid && referenceArguments == 1 {
			return map[string]string{"transform_kind": functionName, "transform_expression": expression}
		}
	}
	if _, ok := dbtMultiInputRowLevelFunctions[functionName]; ok && len(arguments) >= 2 && supportsRowLevelArguments(arguments) {
		return map[string]string{"transform_kind": functionName, "transform_expression": expression}
	}
	if functionName == "concat_ws" && len(arguments) >= 3 && isLiteralExpression(arguments[0]) && supportsRowLevelArguments(arguments[1:]) {
		return map[string]string{"transform_kind": functionName, "transform_expression": expression}
	}
	return nil
}

func supportsRowLevelArguments(arguments []string) bool {
	referenceCount := 0
	for _, argument := range arguments {
		if isSupportedRowLevelExpression(argument) {
			referenceCount++
			continue
		}
		return false
	}
	return referenceCount >= 1
}

func simpleReferenceTokens(expression string) []string {
	matchedIdentifiers := make(map[string]struct{})
	tokens := make([]string, 0)
	seen := make(map[string]struct{})
	for _, match := range qualifiedReferenceMatches(expression) {
		token := match.Alias + "." + match.Column
		matchedIdentifiers[match.Alias] = struct{}{}
		matchedIdentifiers[match.Column] = struct{}{}
		if _, ok := seen[token]; ok {
			continue
		}
		seen[token] = struct{}{}
		tokens = append(tokens, token)
	}
	for _, identifier := range unqualifiedIdentifiers(expression, matchedIdentifiers) {
		if _, ok := seen[identifier]; ok {
			continue
		}
		seen[identifier] = struct{}{}
		tokens = append(tokens, identifier)
	}
	return tokens
}

func hasUnsupportedFunctionCall(expression string) bool {
	for _, match := range dbtFunctionCallScanRe.FindAllStringSubmatch(expression, -1) {
		switch strings.ToLower(strings.TrimSpace(match[1])) {
		case "and", "case", "else", "end", "in", "is", "not", "or", "then", "when":
			continue
		default:
			return true
		}
	}
	return false
}

func collapsedShape(expression string, references []string, stripCaseKeywords bool) string {
	sanitized := dbtSingleQuotedLiteralScan.ReplaceAllString(expression, "0")
	sanitized = dbtNumericLiteralScan.ReplaceAllString(sanitized, "${1}0${3}")
	sanitized = replaceReferenceTokens(sanitized, references)
	if stripCaseKeywords {
		sanitized = dbtCaseKeywordRe.ReplaceAllString(sanitized, " ")
	}
	return strings.ReplaceAll(strings.ReplaceAll(sanitized, "REF", ""), "0", "")
}

func replaceReferenceTokens(expression string, references []string) string {
	sanitized := expression
	for _, token := range references {
		re := regexp.MustCompile(`\b` + regexp.QuoteMeta(token) + `\b`)
		sanitized = re.ReplaceAllString(sanitized, "REF")
	}
	return sanitized
}

func isSimpleReferenceExpression(expression string) bool {
	normalized := stripWrappingParentheses(strings.TrimSpace(expression))
	return normalized != "" && (dbtBareIdentifierRe.MatchString(normalized) || dbtQualifiedReferenceRe.MatchString(normalized))
}

func isLiteralExpression(expression string) bool {
	normalized := stripWrappingParentheses(strings.TrimSpace(expression))
	if normalized == "" {
		return false
	}
	switch strings.ToLower(normalized) {
	case "null", "true", "false":
		return true
	}
	return dbtSingleQuotedLiteralRe.MatchString(normalized) || dbtNumericLiteralRe.MatchString(normalized)
}

func splitTopLevelArguments(arguments string) []string {
	items := make([]string, 0)
	current := make([]rune, 0, len(arguments))
	depth := 0
	inSingleQuote := false
	var prev rune
	for _, character := range arguments {
		if character == '\'' && prev != '\\' {
			inSingleQuote = !inSingleQuote
		} else if !inSingleQuote {
			switch character {
			case '(':
				depth++
			case ')':
				if depth > 0 {
					depth--
				}
			case ',':
				if depth == 0 {
					item := strings.TrimSpace(string(current))
					if item != "" {
						items = append(items, item)
					}
					current = current[:0]
					prev = character
					continue
				}
			}
		}
		current = append(current, character)
		prev = character
	}
	if tail := strings.TrimSpace(string(current)); tail != "" {
		items = append(items, tail)
	}
	return items
}

type qualifiedReferenceMatch struct {
	Alias  string
	Column string
}

func qualifiedReferenceMatches(expression string) []qualifiedReferenceMatch {
	indexes := dbtQualifiedReferenceScanRe.FindAllStringSubmatchIndex(expression, -1)
	matches := make([]qualifiedReferenceMatch, 0, len(indexes))
	for _, indexSet := range indexes {
		if len(indexSet) < 6 {
			continue
		}
		if next := nextNonSpaceCharacter(expression, indexSet[1]); next == "(" {
			continue
		}
		matches = append(matches, qualifiedReferenceMatch{
			Alias:  expression[indexSet[2]:indexSet[3]],
			Column: expression[indexSet[4]:indexSet[5]],
		})
	}
	return matches
}

func splitCastArguments(arguments string) (string, string, bool) {
	depth := 0
	inSingleQuote := false
	lowerArguments := strings.ToLower(arguments)
	for index, character := range arguments {
		if character == '\'' && (index == 0 || arguments[index-1] != '\\') {
			inSingleQuote = !inSingleQuote
			continue
		}
		if inSingleQuote {
			continue
		}
		switch character {
		case '(':
			depth++
			continue
		case ')':
			if depth > 0 {
				depth--
			}
			continue
		}
		if depth != 0 || index+4 > len(arguments) || lowerArguments[index:index+4] != " as " {
			continue
		}
		return strings.TrimSpace(arguments[:index]), strings.TrimSpace(arguments[index+4:]), true
	}
	return "", "", false
}
